### PR TITLE
WT-9268 Delay deletion the history store record to reconciliation

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

**Comment by quchenhao** [comment]
@agorrod @bvpvamsikrishna With this change, we can get rid of the checkpoint_reserved_txn_id (WT-7630). I plan to remove these logic in a following on pr.

File: src/reconcile/rec_visibility.c
Comment (kommiharibabu) [inline src/reconcile/rec_visibility.c:300]: if the stop timestamp is NONE, this record shouldn't be visible right? Do we really need to check for this condition?

File: src/history/hs_rec.c
Comment (kommiharibabu) [inline src/history/hs_rec.c:496]: As we are not inserting the same update again into the history store because it already exists can this lead to a situation where the history store update doesn't have a stop timestamp?

Prepare update P1 -> U1.
Reconcile leads P1 to the data store and U1 to history store with MAX stop timestamp.
Rollback prepare transaction leaves the U1 in the HS.
Insert another update U2
Reconciliation leaves U1 as it is.

Are we handling this scenario somewhere?

Comment (quchenhao) [reply src/history/hs_rec.c:496]: Yes, we are. Now rec_upd_select will walk the whole tree to delete any update with the WT_UPDATE_TO_DELETE_FROM_HS flag. Actually in that case, there is no need to check this flag in hs code.

File: src/reconcile/rec_visibility.c
Comment (quchenhao) [reply src/reconcile/rec_visibility.c:300]: HS record is always visible regardless of tombstone. I'm not sure what you mean.

File: src/history/hs_rec.c
Comment (quchenhao) [reply src/history/hs_rec.c:496]: I think we should see this flag in reconciliation. If checkpoint writes the restored update to the data store, we should not see this in hs code. If it keeps the prepared update, it should not even save the update chain so we can't be here. If it writes a newer update, checkpoint should have removed it from hs in rec_upd_select.

File: src/reconcile/rec_visibility.c
Comment (quchenhao) [reply src/reconcile/rec_visibility.c:300]: Fixed in 3e2bd757bdf59bfccc77b1370ced624cce320626

File: src/reconcile/rec_visibility.c
Comment (agorrod) [inline src/reconcile/rec_visibility.c:809]: The rec_upd_select function doesn't seem like the best candidate for doing this fixup. It's already one of the most complex functions in the system, and that's partly because we've overloaded it's job of selecting which update should be written to the data file.

Could this history store fixup happen later during the reconciliation process?

File: src/reconcile/rec_visibility.c
Comment (agorrod) [inline src/reconcile/rec_visibility.c:252]: This seems to belong in src/history/hs_rec.c - or am I missing something?

File: src/reconcile/rec_visibility.c
Comment (agorrod) [inline src/reconcile/rec_visibility.c:274]: Is there a pattern spread across our code of building a history store key? It seems complex, and layers in knowledge about row/column stores which it's better to keep out of general reconciliation code like this (i.e: I'd rather have fewer places that conditionally check for row vs column store pages.

File: src/reconcile/rec_visibility.c
Comment (agorrod) [inline src/reconcile/rec_visibility.c:822]: The `for, break, if, for` construct here seems odd to me. Is there a guarantee that there is only ever 0, 1 or 2 updates with WT_UPDATE_TO_DELETE_FROM_HS in an update chain, with 0 being the common case. 1 indicating that there is no tombstone (what does that mean?), and 2 indicating that there is a value and a tombstone?

**Review by agorrod — CHANGES_REQUESTED** [review]
I've got a few questions about the code structure and layout. I found some hard to capture in writing, so ping me if you have questions.

File: src/reconcile/rec_visibility.c
Comment (quchenhao) [reply src/reconcile/rec_visibility.c:809]: It can happen later when we inserting the updates to the history store. However, it is also a complex function.

File: src/reconcile/rec_visibility.c
Comment (quchenhao) [reply src/reconcile/rec_visibility.c:252]: We can move it there.

File: src/reconcile/rec_visibility.c
Comment (quchenhao) [reply src/reconcile/rec_visibility.c:274]: There are other places we do the same. I can extract it to a separate function.

File: src/reconcile/rec_visibility.c
Comment (quchenhao) [reply src/reconcile/rec_visibility.c:822]: Yes, that's correct, at most two updates with the flag WT_UPDATE_TO_DELETE_FROM_HS. I'm not sure how this translate to the change of the code structure.

File: src/reconcile/rec_visibility.c
Comment (quchenhao) [reply src/reconcile/rec_visibility.c]: Fixed in [ffea256](https://github.com/wiredtiger/wiredtiger/pull/7939/commits/ffea25641f7d62ac1919f643ba0d12ae1b3c2075)

File: src/reconcile/rec_visibility.c
Comment (quchenhao) [reply src/reconcile/rec_visibility.c]: Fixed in [ffea256](https://github.com/wiredtiger/wiredtiger/pull/7939/commits/ffea25641f7d62ac1919f643ba0d12ae1b3c2075)

File: src/reconcile/rec_visibility.c
Comment (quchenhao) [reply src/reconcile/rec_visibility.c]: This is not needed anymore after I moved the code to hs_rec.c

File: src/reconcile/rec_visibility.c
Comment (quchenhao) [reply src/reconcile/rec_visibility.c]: Fixed in [ffea256](https://github.com/wiredtiger/wiredtiger/pull/7939/commits/ffea25641f7d62ac1919f643ba0d12ae1b3c2075)

**Review by agorrod — COMMENTED** [review]
I don't have time for a full review today, but a high level review, this looks much better to me.

Perhaps someone else can do a more detailed review in the meantime, to avoid this getting blocked on me?

File: src/history/hs_rec.c
Comment (kommiharibabu) [inline src/history/hs_rec.c:84]: Can you add an assert here to check that this history store entry has its own tombstone? Otherwise, the entry shouldn't be removed.

File: src/txn/txn.c
Comment (kommiharibabu) [inline src/txn/txn.c:1363]: This function is called only for `commit` now. You can add a check outside and change passing the parameters.

File: src/reconcile/rec_visibility.c
Comment (kommiharibabu) [inline src/reconcile/rec_visibility.c:837]: In this scenario, the same update is present in both data and history store. Moving this update again to the history store can trigger the OOO fix.

File: test/suite/test_prepare21.py
Comment (kommiharibabu) [inline test/suite/test_prepare21.py:109]: Now we don't need a parallel running checkpoint thread, by default rollback we don't remove the entry. Remove this thread logic? I know that the next one is required.

File: src/history/hs_rec.c
Comment (quchenhao) [reply src/history/hs_rec.c:84]: I can't as it returns not found so the cursor is reset.

File: src/reconcile/rec_visibility.c
Comment (quchenhao) [reply src/reconcile/rec_visibility.c:837]: No before we move it, we first delete it from the history store.

Comment (quchenhao) [reply src/reconcile/rec_visibility.c:837]: No, it has the hs flag so will not be moved to hs again.

File: src/history/hs_rec.c
Comment (kommiharibabu) [reply src/history/hs_rec.c:84]: I mean adding an assert to check that `delete_tombstone != NULL`

File: src/reconcile/rec_visibility.c
Comment (kommiharibabu) [reply src/reconcile/rec_visibility.c:837]: Yeah, understood, even in rollback, we set the flag to delete them before reinsert.

File: src/txn/txn.c
Comment (quchenhao) [reply src/txn/txn.c]: Done

File: src/history/hs_rec.c
Comment (quchenhao) [reply src/history/hs_rec.c:85]: Done

File: test/suite/test_prepare21.py
Comment (kommiharibabu) [inline test/suite/test_prepare21.py:111]: Something wrong happened that we are not inserting `value_d` to trigger and check the OOO during the reorganisation of the test.

**Review by kommiharibabu — APPROVED** [review]
LGTM

File: test/suite/test_prepare21.py
Comment (quchenhao) [reply test/suite/test_prepare21.py:113]: Sorry, fixed.

**Comment by agorrod** [comment]
@quchenhao is this waiting on review from me? Do the test failures need investigation?

**Comment by quchenhao** [comment]
@agorrod I'm still investigating one failure. Let me first address that.

**Comment by quchenhao** [comment]
@agorrod @bvpvamsikrishna This is ready to review now.

**Review by agorrod — APPROVED** [review]
Again, I haven't done a detailed review, but the structure and intent looks good. Thanks for being patient with my suggested changes.
If you can walk someone else through the changes in the pr or get a review from Vamsi I think this is ready to go

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm.
As discussed, my previous comment can be ignored as we handle if we don't find the entry in history store.

**Pull request merged**
